core/trino-parser/src/test/java/io/trino/sql/parser/TestSqlParser.java

```diff
@@ -3382,85 +3376,123 @@ public void testDeny()
         assertStatement("DENY INSERT, DELETE ON t TO u",
                 new Deny(
                         Optional.of(ImmutableList.of("INSERT", "DELETE")),
-                        Optional.empty(),
-                        QualifiedName.of("t"),
+                        new GrantObject(location(1, 1), Optional.empty(), QualifiedName.of("t")),
                         new PrincipalSpecification(PrincipalSpecification.Type.UNSPECIFIED, new Identifier("u"))));
         assertStatement("DENY UPDATE ON t TO u",
                 new Deny(
                         Optional.of(ImmutableList.of("UPDATE")),
-                        Optional.empty(),
-                        QualifiedName.of("t"),
+                        new GrantObject(location(1, 1), Optional.empty(), QualifiedName.of("t")),
                         new PrincipalSpecification(PrincipalSpecification.Type.UNSPECIFIED, new Identifier("u"))));
         assertStatement("DENY ALL PRIVILEGES ON TABLE t TO USER u",
                 new Deny(
                         Optional.empty(),
-                        Optional.of(GrantOnType.TABLE),
-                        QualifiedName.of("t"),
+                        new GrantObject(location(1, 1), Optional.of("TABLE"), QualifiedName.of("t")),
                         new PrincipalSpecification(PrincipalSpecification.Type.USER, new Identifier("u"))));
         assertStatement("DENY SELECT ON SCHEMA s TO USER u",
                 new Deny(
                         Optional.of(ImmutableList.of("SELECT")),
-                        Optional.of(GrantOnType.SCHEMA),
-                        QualifiedName.of("s"),
+                        new GrantObject(location(1, 1), Optional.of("SCHEMA"), QualifiedName.of("s")),
                         new PrincipalSpecification(PrincipalSpecification.Type.USER, new Identifier("u"))));
     }
 
+    @Test
+    public void testExoticPrivilegesAndEntityKinds()
+    {
+        for (String privilege : ImmutableSet.of("SELECT", "INSERT", "AUTO_GYRATE")) {
+            for (String entityKind : ImmutableSet.of("TABLE", "SCHEMA", "FUNKY_ENTITY")) {
+                for (Type type : Type.values()) {
+                    testExoticPrivilegeGrants(privilege, Optional.of(entityKind), type);
+                    testExoticPrivilegeGrants(privilege, Optional.empty(), type);
+                }
+            }
+        }
+
+        // Failing cases
+        // TODO: Add more failing cases
+        assertThatThrownBy(() -> SQL_PARSER.createExpression("GRANT select, role23 TO role123"))
+                .isInstanceOf(ParsingException.class)
+                .hasMessageMatching("line 1:7: mismatched input 'select'. Expecting:.*");
+    }
+
+    private void testExoticPrivilegeGrants(String privilege, Optional<String> entityKind, PrincipalSpecification.Type type)
+    {
+        assertStatement(formatGrant("GRANT %s ON %st TO %su", privilege, entityKind, type),
```

Comment:
Unfortunately, `assertThat(statement(...))` requires accurate determination of the location of element, but these are tests whose SQL strings are constructed, so it is nontrivial to come up with the exact locations.

I just checked, and there are 352 uses of `assertStatement` and 81 uses of `assertThat(statement(...))`.  Unless you can think of a good way to come up with the locations, it seems better to me to allow `assertStatement` here.

Reply:
We're trying to get rid of `assertStatement`. If we keep adding test cases that use it, we just make the migration harder for whoever ends up doing it. Why not do it the right way in the first place?

> requires accurate determination of the location of element, but these are tests whose SQL strings are constructed, so it is nontrivial to come up with the exact locations.

Then we should structure the tests in a different way so that we don't run into that problem. Let me take a closer look at what they are doing.

Reply:
I removed the generated tests, and wrote out the important tests long-hand using `assertThat(statement(...)).isEqualTo(...)`.